### PR TITLE
feat(connectors): G3.7-T7 HetznerRobotConnector skeleton — HttpConnector + HTTP Basic + no-retry-on-401 + _post_form + fingerprint + probe + registry v2

### LINKED ISSUE #846: Task: G3.7-T7 HetznerRobotConnector skeleton — HttpConnector + HTTP Basic (Webservice user) + no-retry-on-401 + form-encoded helper + fingerprint + probe + registry v2
Parent initiative: #370
Parent goal: #214
Depends-on: (none — G0.2-T3 HttpConnector #242 closed)

## Target

- Add: `backend/src/meho_backplane/connectors/hetzner_robot/{__init__.py, connector.py, session.py}`
- Add: `backend/tests/test_connectors_hetzner_robot_auth.py` (respx)
- Precedent: [`connectors/harbor/connector.py`](../blob/main/backend/src/meho_backplane/connectors/harbor/connector.py) (HTTP Basic + injectable Vault loader + fingerprint/probe/execute), [`connectors/adapters/http.py`](../blob/main/backend/src/meho_backplane/connectors/adapters/http.py) (`HttpConnector` L50).

## Context

Hetzner Robot is a **generic-ingested** HTTP connector per #370. Inherits `HttpConnector`. Auth is **HTTP Basic** against the Hetzner **Webservice user** (distinct from the Robot login user). **CRITICAL:** 3 failed 401s from one source IP trigger a 10-minute IP block (consumer-wrapper-documented) — so the connector MUST **NOT retry on 401**; it fails fast with `auth_failed` + an instructional message pointing at the credential, protecting every operator on the shared egress IP. The Robot API requires `application/x-www-form-urlencoded` bodies (NOT JSON) — a `_post_form` helper ships now for v0.2.next write readiness (v0.2 read ops never POST). This Task ships the skeleton; ops via G0.7 ingestion in T8.

## Current state

No `connectors/hetzner_robot/` package. Hetzner Robot targets unreachable through MEHO; operators call `./scripts/hetzner-robot.sh`.

## Desired state

`HetznerRobotConnector(HttpConnector)`, `product="hetzner-robot"`, `version="2026-04"`, `impl_id="hetzner-rest"`, registered via `register_connector_v2(...)`. Auth: HTTP Basic (Webservice user) from the Vault `secret_ref` via an injectable loader (mirror harbor `_load_credentials` + `_basic_auth_header`). **No-retry-on-401**: a 401 raises `auth_failed` immediately (never re-attempts). `_post_form(target, path, data)` helper sends `application/x-www-form-urlencoded` (httpx `data=`). `auth_headers()` rejects non-`shared_service_account` `auth_model` with `NotImplementedError`. `fingerprint()` → `GET /server` → `vendor="hetzner"`, `product="robot-webservice"`, account_id, server_count. `probe()` → TCP + cert + `GET /server` (401-not-retried). `execute()` = the G0.6 shim (`connector_id="hetzner-rest-2026-04"`).

## Acceptance criteria

- [ ] `connectors/hetzner_robot/` exists; `register_connector_v2(product="hetzner-robot", version="2026-04", impl_id="hetzner-rest", cls=HetznerRobotConnector)`; `test_connectors_registry_v2.py` asserts the triple resolves.
- [ ] respx unit test: Basic auth header; **401 → `auth_failed` raised immediately with instructional message, NO retry (assert exactly one request issued)**; `_post_form` sends `x-www-form-urlencoded`; Vault loader missing key → `RuntimeError`; per-target client isolation; `auth_model="per_user"` → `NotImplementedError`.
- [ ] `fingerprint()` against mocked `GET /server` returns account_id + server_count; unreachable → `reachable=False` + `extras["error"]`.
- [ ] `probe()` returns `ok=True`/`ok=False`+reason; a 401 during probe is not retried.
- [ ] `ruff check` + `mypy` clean; `pytest -n auto backend/tests/test_connectors_hetzner_robot_auth.py backend/tests/test_connectors_registry_v2.py` passes.

## Out of scope

- Hetzner Robot ops — T8 ingestion.
- Writes (server reset, vSwitch mutation, cancellation, rDNS edits) — deferred per #370.
- Hetzner Cloud (the second Hetzner product) — out of scope.
- CLI verbs / MCP review / onboarding doc — T9.

## References

- Precedent: [`connectors/harbor/connector.py`](../blob/main/backend/src/meho_backplane/connectors/harbor/connector.py) (Basic + loader + fingerprint/probe/execute), [`connectors/adapters/http.py`](../blob/main/backend/src/meho_backplane/connectors/adapters/http.py) (`HttpConnector` L50), `register_connector_v2` [`registry.py:108`](../blob/main/backend/src/meho_backplane/connectors/registry.py#L108).
- Pinned (`backend/pyproject.toml`): httpx>=0.27 (0.28.1 resolved; `data=` for form-encoded), tenacity>=9.0, pydantic>=2.13.4, respx>=0.21.
- Hetzner Robot Webservice docs: https://robot.hetzner.com/doc/webservice/en.html
- Consumer wrapper this mirrors: https://github.com/evoila-bosnia/claude-rdc-hetzner-dc/blob/main/scripts/hetzner-robot.sh
- Best-practices anchors (by pack name): **python_best_practices** — httpx pooling, Vault-loader injection, Pydantic v2 schemas; **devops_best_practices** — no-retry-on-auth-fail discipline (shared-egress IP-block protection).

### PR DESCRIPTION
**Pull request by kr3s0**
## Summary

- Adds `HetznerRobotConnector(HttpConnector)` skeleton for the Hetzner Robot Webservice API, registered under `(product="hetzner-robot", version="2026-04", impl_id="hetzner-rest")` via `register_connector_v2`
- Auth: HTTP Basic against the Webservice user (distinct from Robot login user); credentials loaded via injectable Vault loader; cached per-target
- **No-retry-on-401**: a 401 raises `auth_failed` immediately with an instructional message — protects shared-egress operators from Hetzner Robot's 10-min IP block (triggered by 3 consecutive 401s from one source IP)
- `_post_form(target, path, data)` helper sends `application/x-www-form-urlencoded` bodies for v0.2.next write readiness (Robot API requires form-encoded, not JSON)
- `fingerprint()` → `GET /server` → `vendor="hetzner"`, `product="robot-webservice"`, `account_id`, `server_count`; `probe()` → TCP + TLS + `GET /server` (401-not-retried)

## Test plan

- [x] `ruff check` — clean
- [x] `ruff format --check` — clean
- [x] `mypy src/meho_backplane/connectors/hetzner_robot/` — no issues
- [x] `pytest -n auto tests/test_connectors_hetzner_robot_auth.py tests/test_connectors_registry_v2.py` — 27 new hetzner tests pass; 6 failing `test_connectors_registry_v2.py::test_*_connector_registered_under_v2_triple` tests are pre-existing on `main` (confirmed baseline)
- [x] registry triple `("hetzner-robot", "2026-04", "hetzner-rest")` resolves to `HetznerRobotConnector`
- [x] 401 raises `auth_failed` immediately; `route.call_count == 1` (no retry) asserted
- [x] `_post_form` sends `application/x-www-form-urlencoded` Content-Type with form-encoded body
- [x] Vault loader missing key → `RuntimeError` naming target and missing key
- [x] `auth_model="per_user"` → `NotImplementedError` naming target and mode
- [x] `fingerprint()` returns `account_id + server_count`; unreachable → `reachable=False + extras["error"]`
- [x] `probe()` returns `ok=True`/`ok=False+reason`; 401 during probe not retried
- [x] `docs/codebase/connectors-hetzner-robot.md` written

## Out of scope

- Hetzner Robot ops — T8 spec ingestion
- Write ops (server reset, vSwitch, cancellation, rDNS) — deferred per #370
- Hetzner Cloud — separate product, out of scope
- CLI verbs / MCP review / onboarding doc — T9

Closes #846

## Summary by CodeRabbit

* **New Features**
  * Added Hetzner Robot connector: HTTP Basic auth, server fingerprinting, discovery and probing, and per-target credential caching; default credential loader is a placeholder requiring a custom loader to be provided.

* **Documentation**
  * Added connector docs covering configuration, auth behavior, probe/fingerprint semantics, and write-body requirements.

* **Tests**
  * Added tests covering auth behavior, credential caching, probe/fingerprint parsing, registration, and client lifecycle.

[Review Change Stack](https://app.coderabbit.ai/change-stack/evoila/meho/pull/906?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by kr3s0** [comment]
@coderabbitai review

**Review by kr3s0 — COMMENTED** [review]
## /auto-review-pr summary (iteration 1)




**Anchor task:** #846 — G3.7-T7 HetznerRobotConnector skeleton — HttpConnector + HTTP Basic (Webservice user) + no-retry-on-401 + form-encoded helper + fingerprint + probe + registry v2
**Surface area:** backend
**Best-practices packs audited:** python
**Verdict:** REQUEST_CHANGES
**Head SHA reviewed:** `6b7ed8a5c9bd5a2834efcde0cd2503f888a966fa`
**AI reviewer (CodeRabbit):** received in 272s (09:30:21Z → 09:34:53Z)

### Acceptance criteria
- [x] `connectors/hetzner_robot/` exists; `register_connector_v2(product="hetzner-robot", version="2026-04", impl_id="hetzner-rest", cls=HetznerRobotConnector)`; `test_connectors_registry_v2.py` asserts the triple resolves. — Delivered; 27 hetzner tests + new registry test pass locally.
- [x] respx unit test: Basic auth header; 401 → `auth_failed` raised immediately with instructional message, NO retry; `_post_form` sends `x-www-form-urlencoded`; Vault loader missing key → `RuntimeError`; per-target client isolation; `auth_model="per_user"` → `NotImplementedError`. — All 27 tests pass.
- [x] `fingerprint()` against mocked `GET /server` returns account_id + server_count; unreachable → `reachable=False` + `extras["error"]`. — Verified by tests.
- [x] `probe()` returns `ok=True`/`ok=False`+reason; a 401 during probe is not retried. — Verified by tests.
- [❌] `ruff check` + `mypy` clean; `pytest -n auto backend/tests/test_connectors_hetzner_robot_auth.py backend/tests/test_connectors_registry_v2.py` passes. — The task's stated tests pass, but `test_api_v1_connectors_ingest.py::test_register_connector_v2_round_trip_lossless_for_every_entry` fails in CI with `AssertionError: version round-trip lost for ('hetzner-robot', '2026-04', 'hetzner-rest')`.

### Project-board alignment
| Issue | Status | Link | Auto-close | Delivered | Action |
|---|---|---|---|---|---|
| #846 | Open | Closes | ✅ | ⚠️ (partial — CI red) | close on merge after CI fixed |

### Predicted vs. actual diff
- ✅ `backend/src/meho_backplane/connectors/hetzner_robot/__init__.py` — expected, correct registration pattern
- ✅ `backend/src/meho_backplane/connectors/hetzner_robot/connector.py` — expected, HttpConnector subclass with auth/fingerprint/probe/execute
- ✅ `backend/src/meho_backplane/connectors/hetzner_robot/session.py` — expected, credential Protocols + loader stub
- ✅ `backend/tests/test_connectors_hetzner_robot_auth.py` — expected, 27 respx tests covering all acceptance criteria
- ✅ `backend/tests/test_connectors_registry_v2.py` — expected addition of Hetzner Robot triple test
- ✅ `docs/codebase/connectors-hetzner-robot.md` — expected, docs shipped per task checklist
- 🕳️ `backend/tests/test_api_v1_connectors_ingest.py` — missing update to `_every_v2_connector_registered` fixture and the `"2026-04"` version string in the fixture's hardcoded connector list must be consistent with a parseable format

### Integration verification
- [x] `ruff check src/` — clean (local)
- [x] `ruff format --check src/` — 228 files already formatted (local)
- [x] `mypy src/meho_backplane/connectors/hetzner_robot/ --ignore-missing-imports` — clean (local)
- [x] `pytest tests/test_connectors_hetzner_robot_auth.py` — 27 passed (local)
- [❌] `pytest tests/test_api_v1_connectors_ingest.py::test_register_connector_v2_round_trip_lossless_for_every_entry` — **FAILS in CI** (xdist worker import collision causes hetzner triple to appear in registry with unparseable hyphenated version)
- [x] `pytest tests/test_connectors_registry_v2.py` — new test passes (the 6 pre-existing failures are on main)
- [⚠️ not run] migrations replay — n/a (no migrations in this PR)
- [⚠️ not run] docker compose up — not required for connector skeleton

### Findings (inline — each is a merge blocker per team policy)
For the `/auto-implement-issue --continue-pr` punch list:

- 🛑 B1 — `version="2026-04"` fails parse_connector_id round-trip; CI is red; required change: change to `"2026.04"` everywhere AND add HetznerRobotConnector to the `_every_v2_connector_registered` fixture
- ⚠️ M1 — `_every_v2_connector_registered` fixture in `test_api_v1_connectors_ingest.py` not updated with the new connector; required change: add entry to fixture (depends on B1 fix)
- ⚠️ M2 — `resp.json()` called without `try/except ValueError` in `_get_robot_json` (line 237) and `_post_form` (line 265); required change: wrap each `resp.json()` call with `try/except (ValueError, Exception)` and raise a contextual `RuntimeError` naming target+path

### Adjacent improvements (non-blocking, not posted inline)
- The docstring coverage is 72.92% vs the configured 80% threshold (CodeRabbit pre-merge check). Some private helper methods lack docstrings (`_is_acceptable_auth_model`, `_basic_auth_header`, `_get_robot_json`, `_post_form`). Fixing this would address the CodeRabbit warning but is not a CI gate in the test suite itself.
- The `execute()` shim's comment says `connector_id = "hetzner-rest-2026-04"` — this will need updating if the version format changes to `"2026.04"`.

### AI reviewer (CodeRabbit) cross-check
- Adopted: 1 (M2 — `resp.json()` without `ValueError` guard, cited inline)
- Disputed: 0
- Out of scope: 0

### Docs
- [x] `docs/codebase/connectors-hetzner-robot.md` — present, accurate for the skeleton; will need a minor version string update if B1 fix changes the version format

### What worked well
- No-retry-on-401 implementation is correct: exactly one request is made, the `auth_failed` message is instructional, the shared-egress IP-block risk is documented prominently throughout. The test asserting `route.call_count == 1` is excellent coverage for the most operationally critical requirement.
- The credential injection pattern with an async `Callable` loader mirrors the Harbor precedent cleanly; per-target cache isolation under a single `asyncio.Lock` is correct.
- 553-line test file has thorough coverage of the auth boundary conditions. The `_clean_hetzner_robot_registry` fixture pattern correctly mirrors the harbor auth test's approach.

File: backend/src/meho_backplane/connectors/hetzner_robot/connector.py [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+    this connector raises :exc:`RuntimeError` with an instructional message
+    naming the target and the Vault path to check.
+
+    The :attr:`priority` is set to ``1`` so a future ``GenericRestConnector``
+    auto-shim (priority=0) loses the registry tie-break if both register for
+    the same triple.
+    """
+
+    product = "hetzner-robot"
+    version = "2026-04"
```
Comment (kr3s0) [inline backend/src/meho_backplane/connectors/hetzner_robot/connector.py]: **[🛑 B1]** — `version="2026-04"` fails `parse_connector_id` round-trip; CI red

**What:** `build_connector_id("hetzner-robot", "2026-04", "hetzner-rest")` produces `"hetzner-rest-2026-04"`. `parse_connector_id` uses `_CONNECTOR_ID_TAIL_VERSION = re.compile(r"^(?P<head>.+)-(?P<version>[0-9][A-Za-z0-9._]*)$")` — the version charset `[A-Za-z0-9._]` excludes `-`, so the regex matches `head="hetzner-rest-2026"` and `version="04"` (not `"2026-04"`). CI confirms: `FAILED tests/test_api_v1_connectors_ingest.py::test_register_connector_v2_round_trip_lossless_for_every_entry — AssertionError: version round-trip lost for (hetzner-robot, 2026-04, hetzner-rest): build=hetzner-rest-2026-04 parsed_version=04`.

**Why it matters:** The connector_id round-trip is a hard contract enforced by an existing gate test (`test_register_connector_v2_round_trip_lossless_for_every_entry`). A broken round-trip means the dispatcher cannot resolve ops ingested against this connector from a `connector_id` string — T8 ops ingestion (the next task in this chain) would be silently unfindable.

**Required change:** Change `version = "2026-04"` to `version = "2026.04"` here and in `__init__.py` (line 33), `execute()` docstring (line 356), `session.py` docstrings, `docs/codebase/connectors-hetzner-robot.md`, and the new test in `test_connectors_registry_v2.py`. Use `"2026.04"` (dot separator) — this round-trips correctly through the existing parser without any change to `_lookup.py`. Also update the `_every_v2_connector_registered` fixture in `test_api_v1_connectors_ingest.py` to include `HetznerRobotConnector` with the corrected version (see M1 inline comment on that file).

**Source:** `backend/src/meho_backplane/operations/_lookup.py:48` — `_CONNECTOR_ID_TAIL_VERSION = re.compile(r"^(?P<head>.+)-(?P<version>[0-9][A-Za-z0-9._]*)$")`. All other connectors use dot-separated versions (`"9.0"`, `"2.x"`, `"4.2"`, `"1.x"`, `"9.x"`).

File: backend/tests/test_connectors_registry_v2.py [hunk 2]
```diff
@@ -382,3 +382,24 @@ def test_vcf_fleet_connector_registered_under_v2_triple() -> None:
...
+    register_connector_v2(
+        product=HetznerRobotConnector.product,
+        version=HetznerRobotConnector.version,
+        impl_id=HetznerRobotConnector.impl_id,
+        cls=HetznerRobotConnector,
+    )
+    snapshot = all_connectors_v2()
+    key = ("hetzner-robot", "2026-04", "hetzner-rest")
+    assert key in snapshot
+    assert snapshot[key] is HetznerRobotConnector
```
Comment (kr3s0) [inline backend/tests/test_connectors_registry_v2.py:405]: **[⚠️ M1]** — `_every_v2_connector_registered` fixture in `test_api_v1_connectors_ingest.py` not updated

**What:** The `_every_v2_connector_registered` fixture hardcodes every production v2 connector; `HetznerRobotConnector` is missing. With xdist `-n 6`, the import from `test_connectors_hetzner_robot_auth.py` bleeds into the same worker, the new triple appears in the registry, and `test_register_connector_v2_round_trip_lossless_for_every_entry` fails because version `"2026-04"` does not round-trip. This is part of the same B1 fix: once version is corrected to `"2026.04"`, the new entry must also be added to the fixture.

**Why it matters:** The round-trip test is the gate that catches exactly this class of parser incompatibility. The fixture must include every v2 connector.

**Required change:** In `backend/tests/test_api_v1_connectors_ingest.py`, add to `_every_v2_connector_registered`:
```python
from meho_backplane.connectors.hetzner_robot.connector import HetznerRobotConnector
# in entries tuple:
("hetzner-robot", "2026.04", "hetzner-rest", HetznerRobotConnector),
```
(This is a change to a file not in this PR diff — include it in the fix-up commit.)

File: backend/src/meho_backplane/connectors/hetzner_robot/connector.py [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+        client = await self._http_client(target)
+        headers = await self.auth_headers(target, raw_jwt="")
+        resp = await client.request("GET", path, headers=headers)
+        if resp.status_code == 401:
+            raise RuntimeError(
+                f"auth_failed: Hetzner Robot returned 401 for target "
+                f"{target.name!r} (host={target.host!r}). {_AUTH_FAILED_HINT}"
+            )
+        resp.raise_for_status()
+        return resp.json()
```
Comment (kr3s0) [inline backend/src/meho_backplane/connectors/hetzner_robot/connector.py]: **[⚠️ M2]** — `resp.json()` unguarded against `ValueError` / empty body (also flagged by CodeRabbit)

**What:** `_get_robot_json` at line 237 and `_post_form` at line 265 call `resp.json()` directly after `resp.raise_for_status()`. If the server returns a 200/201 with an empty body or a non-JSON content type, `resp.json()` raises `json.JSONDecodeError` (a `ValueError` subclass) and escapes the helper without being shaped into the `RuntimeError` contract the rest of the connector uses.

**Why it matters:** The `_post_form` helper is the write-path foundation for T8 ops (robot reset, rDNS, vSwitch mutations). Several Robot API write endpoints return 200/201 with empty bodies or plain-text. An unhandled `JSONDecodeError` from a production write op would surface as an opaque stack trace in operator logs rather than a `protocol_error`-shaped `RuntimeError` naming the target and path.

**Required change:**
```python
# _get_robot_json lines 236-237
resp.raise_for_status()
try:
    return resp.json()
except ValueError as exc:
    raise RuntimeError(
        f"protocol_error: Hetzner Robot returned non-JSON response for "
        f"GET {path!r} on target {target.name!r}: {exc}"
    ) from exc
```
Apply the same pattern to `_post_form` (line 265). Return `None` only if the POST helper is explicitly expected to handle 204-no-content responses; otherwise always raise.

**Source:** CodeRabbit review (confirmed valid — `httpx.Response.json()` raises `json.JSONDecodeError` on empty/non-JSON body; `json.JSONDecodeError` is a subclass of `ValueError`). Note: the base `HttpConnector._request_json` and `_post_json` have the same gap — this PR should fix the new code and not regress the existing pattern.

**Review by kr3s0 — COMMENTED** [review]
## /auto-review-pr summary (iteration 2)




**Anchor task:** #846 — G3.7-T7 HetznerRobotConnector skeleton — HttpConnector + HTTP Basic (Webservice user) + no-retry-on-401 + form-encoded helper + fingerprint + probe + registry v2
**Surface area:** backend
**Best-practices packs audited:** python
**Verdict:** APPROVE
**Head SHA reviewed:** `56303eb19924f1d5f4399679cc3c016098607d5f`
**AI reviewer (CodeRabbit):** received in 1s (iter-2 review posted @ 11:35:54Z, reviewed commits 6b7ed8a → 56303eb)

### Acceptance criteria
- [x] `connectors/hetzner_robot/` exists; `register_connector_v2(product="hetzner-robot", version="2026.04", impl_id="hetzner-rest", cls=HetznerRobotConnector)` — verified; version uses dot separator, passes `parse_connector_id` round-trip.
- [x] respx unit test: Basic auth header; 401 → `auth_failed` raised immediately, NO retry; `_post_form` sends `x-www-form-urlencoded`; Vault loader missing key → `RuntimeError`; per-target client isolation; `auth_model="per_user"` → `NotImplementedError` — all 27 hetzner tests pass.
- [x] `fingerprint()` against mocked `GET /server` returns account_id + server_count; unreachable → `reachable=False` + `extras["error"]` — verified by tests.
- [x] `probe()` returns `ok=True`/`ok=False`+reason; 401 during probe not retried — verified by tests.
- [x] `ruff check` + `mypy` clean; `pytest` passes — ruff clean, mypy clean (3 source files), 27 hetzner auth tests pass, `test_register_connector_v2_round_trip_lossless_for_every_entry` passes (was failing in iter-1 due to B1/M1).

### Fixes from iteration 1 — verified resolved
- **B1 (version hyphen):** `version="2026.04"` confirmed in `__init__.py` and `connector.py`. `parse_connector_id("hetzner-rest-2026.04")` → `('hetzner', '2026.04', 'hetzner-rest')` ✅
- **M1 (fixture missing):** `_every_v2_connector_registered` in `test_api_v1_connectors_ingest.py` now includes `("hetzner-robot", "2026.04", "hetzner-rest", HetznerRobotConnector)`. `test_register_connector_v2_round_trip_lossless_for_every_entry` passes ✅
- **M2 (JSON decode guard):** Both `_get_robot_json` (line 237–242) and `_post_form` (line 270–275) have `try/except ValueError` guards that raise `RuntimeError` with context ✅

### Project-board alignment
| Issue | Status | Link | Auto-close | Delivered | Action |
|---|---|---|---|---|---|
| #846 | Open | Closes | ✅ | ✅ (full — all AC met, CI green) | close on merge |

### Predicted vs. actual diff
- ✅ `backend/src/meho_backplane/connectors/hetzner_robot/__init__.py` — correct; version="2026.04"
- ✅ `backend/src/meho_backplane/connectors/hetzner_robot/connector.py` — correct; JSON guards added, version fixed
- ✅ `backend/src/meho_backplane/connectors/hetzner_robot/session.py` — correct
- ✅ `backend/tests/test_connectors_hetzner_robot_auth.py` — correct; 27 respx tests covering all AC
- ✅ `backend/tests/test_connectors_registry_v2.py` — correct; new triple test passes under xdist
- ✅ `backend/tests/test_api_v1_connectors_ingest.py` — correct; fixture and round-trip test updated
- ✅ `docs/codebase/connectors-hetzner-robot.md` — present; one minor doc inaccuracy (see Adjacent improvements)

### Integration verification
- [x] `ruff check` — clean (all new connector files)
- [x] `ruff format --check` — 4 files already formatted
- [x] `mypy` — no issues (3 source files)
- [x] `pytest tests/test_connectors_hetzner_robot_auth.py` — 27 passed
- [x] `pytest tests/test_api_v1_connectors_ingest.py` — 38 passed (including `test_register_connector_v2_round_trip_lossless_for_every_entry`)
- [x] `pytest tests/test_connectors_hetzner_robot_auth.py tests/test_connectors_registry_v2.py -n auto` — 41 passed; 6 pre-existing failures on `test_connectors_registry_v2.py` connector-already-registered tests are identical to main branch baseline (confirmed by running same test suite on main: same 6 failures, same pattern)
- [x] CI: all checks `COMPLETED / SUCCESS` — Python (ruff+mypy+pytest), Python (integration testcontainers), Docker image build, Go, Helm, CodeQL, TruffleHog, Semgrep, DCO ✅
- ⚠️ not run: migrations replay, docker compose up (no migrations in this PR), agent eval (skeleton only — no operations)

### Findings (inline — each is a merge blocker per team policy)
No blocking or major findings in this iteration. All iter-1 blockers resolved.

### Adjacent improvements (non-blocking, not posted inline)
- Module docstring at lines 53–54 says "account ID extracted from the first server's `server_ip` field" but the implementation (line 318) uses `server_number` (with `id` fallback). This is a documentation inaccuracy; the intent matches the code, but a reader of the docstring expecting `server_ip` would be misled. Flagged by CodeRabbit iter-2 as a minor quick-win. Propose fixing in T8 or a follow-on doc pass.
- `_get_robot_json` and `_post_form` JSON decode error messages omit `target.name`/`target.host`. The 401 error messages include target context; the decode errors don't. Not blocking — the path and status code are present. CodeRabbit flagged as `💤 Low value` optional enhancement.

### AI reviewer (CodeRabbit) cross-check
- Adopted: 0 (iter-2 findings are nitpicks/optional; documented in adjacent improvements)
- Disputed: 0
- Out of scope: 0
- Note: CodeRabbit iter-1 finding (JSON decode guard M2) — addressed in this iteration ✅; CodeRabbit iter-2 docstring nit demoted to adjacent improvement (non-blocking) per severity ladder

### Docs
- [x] `docs/codebase/connectors-hetzner-robot.md` — present and accurate, with one minor docstring inaccuracy in `connector.py` module-level docstring (not in the docs/ file; see adjacent improvements)

### What worked well
- The three-punch fix (version dot, JSON guard, fixture entry) addresses the root causes precisely. The `parse_connector_id` round-trip test passing after the hyphen → dot fix demonstrates the fix is correct end-to-end.
- The autouse `_clean_hetzner_robot_registry` fixture in `test_connectors_hetzner_robot_auth.py` correctly mirrors the harbor pattern, ensuring xdist worker isolation.
- 27 respx tests provide thorough coverage of the acceptance criteria including edge cases (per-target isolation, `auth_model=None` backward-compat, enum vs string auth_model).

**Review by zdamir — APPROVED** [review]
